nixos/modules/services/databases/postgresql.nix

```diff
@@ -164,16 +201,46 @@ in
 
   config = mkIf config.services.postgresql.enable {
 
-    services.postgresql.package =
+    assertions =
+      [ # The user specified both 'package' and 'packages', which are mutually exclusive
+        { assertion = (cfg.package != null -> cfg.packages != null);
+          message = ''
+            The option services.postgresql.{package,packages} cannot both be set. Please use the 'packages' option
+            to specify the entire set of PostgreSQL packages.
+          '';
+        }
+
+        # The user specified the old extraPlugins attribute, but they *also* specified some 'plugins' function
+        # which returned a non-empty list. These are mutually exclusive.
+        { assertion = (cfg.extraPlugins != [] -> packageSetPlugins == []);
+          message = ''
+            The option services.postgresql.extraPlugins and .plugins cannot both be non-empty at the same time; please remove
+            uses of .extraPlugins.
+          '';
+        }
+      ];
+
+    warnings =
+      (optional (cfg.package != null)
+        ''The services.postgresql.package attribute is deprecated. Please use services.postgresql.packages (with an 's') instead.
+          See the NixOS manual ('nixos-help') for more information.'') ++
+      (optional (cfg.extraPlugins != [])
+        ''The services.postgresql.extraPlugins attribute is deprecated. Please use services.postgresql.plugins instead,
+          which will correctly override PostgreSQL attributes for you and will provide a list of compatible plugins for
+          the given server version.
+        '');
+
+    services.postgresql.packages =
       # Note: when changing the default, make it conditional on
       # ‘system.stateVersion’ to maintain compatibility with existing
       # systems!
-      mkDefault (if versionAtLeast config.system.stateVersion "17.09" then pkgs.postgresql96
-            else if versionAtLeast config.system.stateVersion "16.03" then pkgs.postgresql95
-            else pkgs.postgresql94);
+      mkDefault (if versionAtLeast config.system.stateVersion "18.09" then pkgs.postgresql10Packages
+            else if versionAtLeast config.system.stateVersion "17.09" then pkgs.postgresql96Packages
+            else if versionAtLeast config.system.stateVersion "16.03" then pkgs.postgresql95Packages
+            else pkgs.postgresql94Packages);
```

Comment:
Can this be made dependent on `cfg.package.version` instead of `config.system.stateVersion`?

Reply:
The intent is that e.g. if you were on NixOS 16.03, and you upgrade to NixOS 17.09, you should keep using PostgreSQL 9.5, because 9.6 is not on-disk compatible, and breaking your database across an upgrade isnt friendly. This is because when you installed NixOS, the (16.03) installer explicitly sets `stateVersion = "16.03"` in `configuration.nix` by default, and *that should not be changed* even beyond upgrades.

I might be misunderstanding what you mean, of course (a diff would be helpful if that's the case!)

Reply:
This isn't a direct problem by itself, but rather a simplification for those who use PG older than NixOS default. When you set `services.postgresql.package` to some old version, you now have to set `services.postgresql.packages` to an older pacakgeset too, which, I think, can be detected by NixOS automatically.